WIP: animate border using JavaScript intervals

webmatze opened the pull request:
This replaces the CSS animation with JavaScript animation.
- should consume less CPU
- gradients look the same as without animation
- gradients are rotated instead of being moved in one dimension

Xapphire13 commented on index.js:
```diff
@@ -25,6 +25,20 @@ module.exports.onRendererWindow = async (window) => {
+  var config = window.config.getConfig();  
+  var deg = 0;
+
+  const updateBorderAngle = () => {
```
Won't doing this in JS be more of a perf issue than letting chromium take care of it?

webmatze replied:
in my system using CSS animations uses double the CPU power than using JavaScript. Sadly.

Xapphire13 replied:
=[. that sucks

webmatze replied:
Yeah. tried it on MacOS and Ubuntu. Furthermore, it is not really possible to animate gradients using CSS animations at the moment, so we had to use hacks that did not look that well.
So unless we find some cool CSS tricks, we should do it with JavaScript.

Xapphire13 commented on index.js:
```diff
@@ -25,6 +25,20 @@ module.exports.onRendererWindow = async (window) => {
+  }
+
+  if (config.hyperBorder && config.hyperBorder.animate) {
+    console.log('start animation interval');
```
Did you mean to leave the `console.log()` line?

webmatze replied:
no. I will remove it.

Xapphire13 commented on index.js:
```diff
@@ -25,6 +25,19 @@ module.exports.onRendererWindow = async (window) => {
     window.document.documentElement.classList.add('elevated');
   }
+
+  var config = window.config.getConfig();  
```
`let` is much safer than using `var` [article on the subject](https://medium.com/craft-academy/javascript-variables-should-you-use-let-var-or-const-394f7645c88f)

I also think your `config` can use `const`

Xapphire13 replied:
We should probable add [no-var](https://eslint.org/docs/rules/no-var) and [prefer-const](https://eslint.org/docs/rules/prefer-const) to the `eslint.json` config

webmatze commented:
I added those two eslint rules and fixed some lint errors. Also the `eslint **/*.js` call in `npm test` did not work correctly before. It only tested js files in sub directories.

Xapphire13 commented:
I wonder if your shell was incorrectly expanding the glob pattern [see eslint docs](https://eslint.org/docs/user-guide/command-line-interface). I think it should be `"**/*.js"` (with the quotes) so that eslint gets the actual glob pattern.

Also why the `-s` on `npm run -s lint`?

webmatze commented:
The `-s` is to [silence additional error messages](https://stackoverflow.com/questions/34426332/how-to-suppress-output-when-running-npm-scripts) from `npm` that you get if the tests are not successful.

When I try the glob pattern from my project root under Ubuntu I get the following results:

```bash
webmatze@webmatze-ThinkPad-25:~/Workplace/github.com/webmatze/hyperborder$ ls **/*.js
test/index.test.js
```

So it definitely misses the `index.js` file in the project root.

Xapphire13 commented:
Your solution doesn't work on Windows:
![image](https://user-images.githubusercontent.com/3392349/32912059-cc031a40-cac2-11e7-9a26-de04f5fe2fc6.png)

The original issue is because the glob pattern wasn't enclosed in quotes (as mentioned in the [eslint docs](https://eslint.org/docs/user-guide/command-line-interface)), so your shell was attempting to expand the glob. And by default bash doesn't expand globstars ([how to enable it](https://makandracards.com/makandra/43638-enabling-in-your-bash)). If you pass the path as a string to ESLint, it will evaluate it as a glob using [minimatch](https://github.com/isaacs/minimatch) and will work on all systems.

Here is an example, I remove some semicolons from both the test file, and the index.js:

On Windows:
![image](https://user-images.githubusercontent.com/3392349/32912323-8c4fff52-cac3-11e7-8ca9-552a2e49d048.png)

On Linux:
![image](https://user-images.githubusercontent.com/3392349/32912359-ac72896c-cac3-11e7-8b62-8590fbc1b3bf.png)

Xapphire13 commented on index.js:
```diff
@@ -27,15 +27,36 @@ module.exports.onRendererWindow = async (window) => {
+      }
+    };
+
+    browserWindow.on('close', clearUpdateBorderAngleInterval );
```
These 3 lines have an inconsistent padded space before the closing paren

webmatze replied:
fixed

Xapphire13 commented on index.js:
```diff
@@ -27,15 +27,36 @@ module.exports.onRendererWindow = async (window) => {
-    deg = deg > 360 ? 2 : deg + 2;
-  };
+    let deg = 0;
+    let intervalId = null;
```
Should you use a module level variable rather than a local one? How does this behave when you modify the config at runtime, do we now have 2 intervals running?

webmatze replied:
Okay, I checked this. The problem here is not that the intervals do not get cleared. This works. But we also have to clear the event listeners when the window is reloaded.

webmatze commented:
Okay, I invested some time and read the [hyper documentation](https://hyper.is/#extensions-api) a little more intensive.
I think we need to do this in a more React compatible way by using `decorateHyper` and return a React component.
Then it should be easier to handle creating and removing event listeners and intervals.

Xapphire13 commented on index.js:
```diff
@@ -25,12 +25,46 @@ module.exports.onRendererWindow = async (window) => {
   if (await isElevated()) {
```
This should be the last thing in the method (other than things that depend on it), since this actually stops the current synchronous execution of the method on the `await` and resumes the remainder at some point in the future.

webmatze replied:
Yes, I know. But there is a weird problem with the `window.config.getConfig()` call. It is not working if I call it before the `await isElevated()` call.
I don't want to dig deeper into this because it can be fixed easily when we switch to `decorateHyper`. Then the config is provided as `props` to the component.

Xapphire13 replied:
Very odd. But you're right. Once this plugin is a React HoC decorator, then lots of issues will be solved =]

Xapphire13 commented:
> So to finish this pull request I will use onUnload to clear all intervals and event listeners, which should do the job, too. onUnload gets called when plugins are beeing reloaded.

Cool! That way we can easily unsubscribe everything without fear of messing up another plugin. Thanks for working through my concerns! This change looks good =]

webmatze commented:
So I will finally merge it. :smile: 
Thanks, @Xapphire13  for reviewing all the changes.
